### PR TITLE
Improve merge and add merge_all

### PR DESCRIPTION
**Pull request by innovate-invent**
This adds a merge function with additional features.
It behaves slightly different than the original merge function and has been included as "merge2" to maintain backwards compatibility of the library.

### CODE REVIEW

**Comment by daler** [comment]
@innovate-invent thanks so much for this and for considering the backwards compatibility.

Is there a way to add arguments to the existing `merge` to get this behavior? It would be much cleaner to have a single method to use for merging.

Also, would you please add some tests for this to the end of `gffutils/test/test.py`?

By the way, thanks for your other PRs #131 and #133!

**Comment by innovate-invent** [comment]
The issue is that the return type is incompatible. merge2 returns a tuple (merged_record, [component_records, ..]). I don't think that can be reconciled while maintaining backwards compatibility.

**Comment by innovate-invent** [comment]
It took some work but I got the interface to be backwards compatible by making a compromise on how the feature children are returned

**Comment by innovate-invent** [comment]
The CI failed because it is using python2. Python 2 will be at end of life at the end of this year. Do you really want to maintain python2 compatibility?

**Comment by daler** [comment]
Thanks so much for this! I especially like the way you're handling the testing, that's elegant.

I'd prefer not to have type annotations in this PR -- I'd like to do a later pass that removes py2 support and adds type annotations so that their usage is consistent throughout the code.

I removed the type annotations from here, fixed an import, and fixed the path in the test. Tests now pass for py2 and py3.

Would you please change the docstring formatting of your additions to match the rest of the code so the sphinx docs pick it up correctly?

**Comment by daler** [comment]
Thanks, looks good! Anything else you'd like to add before merging?

**Comment by innovate-invent** [comment]
This will fail to work without https://github.com/daler/gffutils/pull/136

**Comment by innovate-invent** [comment]
Actually, it looks like it sucked the commits into this PR too

**Comment by daler** [comment]
Ah, right. I was thinking of merging #136 separately to keep the PRs a little more atomic but wanted to add some tests for that first. I'll comment over there.

**Comment by innovate-invent** [comment]
I see you added a test for backwards compatibility. Are the defaults not the same as the original functionality?

File: gffutils/test/test.py [hunk 1]
```diff
@@ -1230,6 +1230,8 @@ def test_issue_119():
     assert db5._autoincrements == {'gene': 2}
     assert db3._autoincrements == db5._autoincrements
 
+    assert db3._autoincrements == db5._autoincrements
```
Comment (innovate-invent) [inline gffutils/test/test.py]: This is a duplicate of the above line ;)

**Comment by daler** [comment]
Right, current master branch does this when merging all the `seq1` features:

```
seq1    .       sequence_feature        1       29      .       .       .
seq1    .       sequence_feature        35      40      .       .       .
```

this PR does the following:

```
seq1    synthetic1,synthetic2   sequence_feature        1       15      .       .       .       ID=sequence_feature_1
seq1    synthetic1      sequence_feature        10      20      .       +       .       ID=strand_plus1
seq1    synthetic1      sequence_feature        10      20      .       -       .       ID=strand_minus1
seq1    synthetic1      sequence_feature        10      29      .       .       .       ID=sequence_feature_2
seq1    synthetic1      sequence_feature        35      40      .       .       .       ID=no_overlap1
```

The new implementation is so much better, but I'd like to be able to document the equivalent arguments needed to retain backwards compatibility.

I was using the following for quick iteration:

After cloning your fork and the master branch, symlink this PR's `test.py` to the master branch's `test.py`,

```
ln -srf pr130/gffutils/test/test.py gffutils/gffutils/test/test.py
```

then you can quickly test both with:

```
(cd pr130 && nosetests gffutils/test/test.py:test_backwards_compatibility);
(cd gffutils && nosetests gffutils/test/test.py:test_backwards_compatibility);

```

File: gffutils/test/test.py (hunk 1 above)
Comment (daler) [reply gffutils/test/test.py]: Fixed!

**Comment by innovate-invent** [comment]
That is odd, I thought I had the argument defaults set to the previous version behavior.
There may be a larger issue with my code if it is doing this.

**Comment by innovate-invent** [comment]
Actually, my synthetic data should toss and exception on the old implementation if ignore_strand == False

**Comment by daler** [comment]
Closing in favor of #151.

The new merging is so much better, I decided to compromise and make the old function still available as `old_merge()` but used everything here as the default merge moving forward. The only other change from #130 is that I removed the `ignore_strand` kwarg that was previously in there for backwards compatibility (as well as the corresponding test using that kwarg).

@innovate-invent thank you again for this!

**Pull request closed without merging**